### PR TITLE
Security Fix for Prototype Pollution - huntr.dev

### PR DESCRIPTION
**Pull request by huntr-helper**
https://huntr.dev/users/alromh87 has fixed the Prototype Pollution vulnerability 🔨. Think you could fix a vulnerability like this?

Get involved at https://huntr.dev/

Q | A
Version Affected | ALL
Bug Fix | YES
Original Pull Request | https://github.com/418sec/dot-json/pull/1
Vulnerability README | https://github.com/418sec/huntr/blob/master/bounties/npm/dot-json/1/README.md

### 📊 Metadata *

`dot-json` is vulnerable to `Prototype Pollution`.
This package allowing for modification of prototype behavior, which may result in Information Disclosure/DoS/RCE.

### ⚙️ Description *

Prototype Pollution refers to the ability to inject properties into existing JavaScript language construct prototypes, such as objects.
JavaScript allows all Object attributes to be altered, including their magical attributes such as _proto_, constructor and prototype.
An attacker manipulates these attributes to overwrite, or pollute, a JavaScript application object prototype of the base object by injecting other values.
Properties on the Object.prototype are then inherited by all the JavaScript objects through the prototype chain.

### 💻 Technical Description *

Fixed by avoiding setting magical attributes.

### 🐛 Proof of Concept (PoC) *

1. Create the following PoC file:

```js
// poc.js
var dotJson = require("./lib/dot-json")
var myfile = new dotJson('myfile.json');
console.log("Before : " + {}.polluted);
myfile.set("__proto__.polluted", "Yes! Its Polluted").save();
console.log("After : " + {}.polluted);
```

2. Execute the following commands in another terminal:

```bash
npm i dot-json # Install affected module
node poc.js #  Run the PoC
```

3. Check the Output:
```
Before : undefined
After : Yes! Its Polluted
```

Captura de pantalla de 2020-12-12 16-12-32

### 🔥 Proof of Fix (PoF) *

After fix execution will block prototype pollution and polluted will be undefined for new objects

Captura de pantalla de 2020-12-12 16-12-50

### 👍 User Acceptance Testing (UAT)

After fix functionality is unafected

### CODE REVIEW

**Comment by JamieSlome** [comment]
@maikelvl - let me know if you have any questions or thoughts, here to help! 🍰

**Review by fregante — APPROVED** [review]

File: lib/dot-json.js [hunk 1]
```diff
@@ -139,7 +139,11 @@ module.exports = function(file) {
 	this._setValueForKeyPath = function(key_path, value) {
 		var key_array = key_path.replace('..', '').split('.');
 		var tip_key = key_array.pop();
+
+		if(key_array.includes('__proto__') || key_array.includes('constructor') || key_array.includes('prototype'))return;
```
Comment (fregante) [inline lib/dot-json.js:143]:
```suggestion
		if (key_array.includes('__proto__') || key_array.includes('constructor') || key_array.includes('prototype')) {
			return;
		}
```

**Comment by maikelvl** [comment]
Thanks a lot!

**Comment by JamieSlome** [comment]
@maikelvl / @fregante, if you want more security fixes and patches like this in the future, you can let security researchers know that they can win bounties protecting your repository by copying this small code snippet into your README.md:

```[huntr](https://huntr.dev)```

👇 👇 👇

[huntr](https://huntr.dev)

**Pull request merged**